File: kotlin/PlanesAndroid/app/src/main/java/com/planes_multiplayer/single_player_engine/PointInfluenceIterator.kt
package com.planes_multiplayer.single_player_engine

class PointInfluenceIterator(point: Coordinate2D) : VectorIterator<Coordinate2D>() {
    private fun generateList() {
        m_internalList.clear()
        for (i in 0..3) {
            val pl = Plane(m_point.x(), m_point.y(), Orientation.values()[i])
            val ppi = PlanePointIterator(pl)
            while (ppi.hasNext()) {
                val point = ppi.next()
                if (m_internalList.indexOf(point) < 0) m_internalList.add(point)
            }
        }
    }

    private val m_point: Coordinate2D

    init {
        m_point = point.clone() as Coordinate2D
        generateList()
    }
}

File: c_plus_plus/singleround/pointinfluenceiterator.cpp
#include "pointinfluenceiterator.h"
#include "planepointiterator.h"
#include "coordinate2d.h"
#include <algorithm>


PointInfluenceIterator::PointInfluenceIterator(const PlanesCommonTools::Coordinate2D& qp):
    PlanesCommonTools::VectorIterator<PlanesCommonTools::Coordinate2D>(),
    m_point(qp)
{
    generateList();
}

void PointInfluenceIterator::generateList()
{
    m_internalList.clear();

	for (int i = 0; i < 4; i++) {
		Plane pl(m_point.x(), m_point.y(), (Plane::Orientation)i);
		PlanePointIterator ppi(pl);

		while (ppi.hasNext()) {
			PlanesCommonTools::Coordinate2D point = ppi.next();
			if (std::find(m_internalList.begin(), m_internalList.end(), point) == m_internalList.end())
				m_internalList.push_back(point);
		}
	}
}
